File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int leftPtr = 0;
        int rightPtr = 1;
        int maxP = 0;
        int profit = 0;
        while (rightPtr < prices.size()){
            if (prices[leftPtr] < prices[rightPtr]){
                profit = prices[rightPtr] - prices[leftPtr];
                maxP = max(maxP, profit);

            } else {
                leftPtr = rightPtr;
            }
            rightPtr++;
        }
        return maxP;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = {} # key -> (i, buy) value -> maxprofit
        def helper(i, buying):
            if i >= len(prices):
                return 0
            if (i, buying) in dp:
                return dp[(i, buying)]

            if buying:
                buy = helper(i+1, not buying) - prices[i]
                cooldown = helper(i+1, buying)
                dp[(i, buying)] =  max(buy, cooldown)
            else:
                sell = helper(i+2, not buying) + prices[i]
                cooldown = helper(i+1, buying)
                dp[(i, buying)] = max(sell, cooldown)
            return dp[(i, buying)]


        return helper(0, True)